### ISSUE TITLE
Installation problem

### ISSUE THREAD

**Issue by zhengrongbin** [kharchenkolab/pagoda2#39]
Dear developer, 
I posted installation problem and got the response. I appreciate for that, but I still met the installation problem, the log information is shown bellow. There is a warning during installation: warning: directory not found for option '-L/usr/local/gfortran/lib/gcc/x86_64-apple-darwin15/6.1.0'.  I am using Mac OS 10.12.6 system, and I cannot install gfortran-4.8.2. However, if I install the latest gfortran, I can only get: /usr/local/gfortran/lib/gcc/x86_64-apple-darwin16/. Could you help to resolve this version problem? or you have some suggestions. Thanks very much!!

> install_github("hms-dbmi/pagoda2")
Downloading GitHub repo hms-dbmi/pagoda2@master
from URL https://api.github.com/repos/hms-dbmi/pagoda2/zipball/master
Installing pagoda2
'/Library/Frameworks/R.framework/Resources/bin/R' --no-site-file --no-environ  \
  --no-save --no-restore --quiet CMD INSTALL  \
  '/private/var/folders/8k/xcdm75p95tz3nz3r9lykjtvh0000gn/T/RtmppNyY8z/devtools21aa34b5ee7e/hms-dbmi-pagoda2-470f1bd'  \
  --library='/Library/Frameworks/R.framework/Versions/3.4/Resources/library'  \
  --install-tests 

* installing *source* package ‘pagoda2’ ...
** libs
/usr/local/clang4/bin/clang++ -std=gnu++11 -I/Library/Frameworks/R.framework/Resources/include -DNDEBUG  -I"/Library/Frameworks/R.framework/Versions/3.4/Resources/library/Rcpp/include" -I"/Library/Frameworks/R.framework/Versions/3.4/Resources/library/RcppArmadillo/include" -I/usr/local/include  -I"/Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include" -I"../inst/include" -I"/Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/lshkit/include" -fopenmp -fPIC  -Wall -g -O2 -c RcppExports.cpp -o RcppExports.o
In file included from RcppExports.cpp:4:
In file included from ./../inst/include/pagoda2.h:15:
In file included from /Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include/space.h:32:
/Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include/utils.h:53:12: warning: 'WARN' macro redefined [-Wmacro-redefined]
#   define WARN(exp) (FILE_LINE_LINK "WARNING: " exp)
           ^
/Library/Frameworks/R.framework/Resources/include/R_ext/RS.h:60:9: note: previous definition is here
#define WARN                    WARNING(NULL)
        ^
In file included from RcppExports.cpp:4:
In file included from ./../inst/include/pagoda2.h:15:
In file included from /Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include/space.h:32:
/Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include/utils.h:66:9: warning: 'ISNAN' macro redefined [-Wmacro-redefined]
#define ISNAN std::isnan
        ^
/Library/Frameworks/R.framework/Resources/include/R_ext/Arith.h:74:11: note: previous definition is here
#  define ISNAN(x)     R_isnancpp(x)
          ^
2 warnings generated.
/usr/local/clang4/bin/clang++ -std=gnu++11 -I/Library/Frameworks/R.framework/Resources/include -DNDEBUG  -I"/Library/Frameworks/R.framework/Versions/3.4/Resources/library/Rcpp/include" -I"/Library/Frameworks/R.framework/Versions/3.4/Resources/library/RcppArmadillo/include" -I/usr/local/include  -I"/Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include" -I"../inst/include" -I"/Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/lshkit/include" -fopenmp -fPIC  -Wall -g -O2 -c binaryExport.cpp -o binaryExport.o
In file included from binaryExport.cpp:7:
In file included from ../inst/include/pagoda2.h:15:
In file included from /Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include/space.h:32:
/Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include/utils.h:53:12: warning: 'WARN' macro redefined [-Wmacro-redefined]
#   define WARN(exp) (FILE_LINE_LINK "WARNING: " exp)
           ^
/Library/Frameworks/R.framework/Resources/include/R_ext/RS.h:60:9: note: previous definition is here
#define WARN                    WARNING(NULL)
        ^
In file included from binaryExport.cpp:7:
In file included from ../inst/include/pagoda2.h:15:
In file included from /Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include/space.h:32:
/Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include/utils.h:66:9: warning: 'ISNAN' macro redefined [-Wmacro-redefined]
#define ISNAN std::isnan
        ^
/Library/Frameworks/R.framework/Resources/include/R_ext/Arith.h:74:11: note: previous definition is here
#  define ISNAN(x)     R_isnancpp(x)
          ^
2 warnings generated.
/usr/local/clang4/bin/clang++ -std=gnu++11 -I/Library/Frameworks/R.framework/Resources/include -DNDEBUG  -I"/Library/Frameworks/R.framework/Versions/3.4/Resources/library/Rcpp/include" -I"/Library/Frameworks/R.framework/Versions/3.4/Resources/library/RcppArmadillo/include" -I/usr/local/include  -I"/Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include" -I"../inst/include" -I"/Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/lshkit/include" -fopenmp -fPIC  -Wall -g -O2 -c misc2.cpp -o misc2.o
In file included from misc2.cpp:7:
In file included from ../inst/include/pagoda2.h:15:
In file included from /Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include/space.h:32:
/Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include/utils.h:53:12: warning: 'WARN' macro redefined [-Wmacro-redefined]
#   define WARN(exp) (FILE_LINE_LINK "WARNING: " exp)
           ^
/Library/Frameworks/R.framework/Resources/include/R_ext/RS.h:60:9: note: previous definition is here
#define WARN                    WARNING(NULL)
        ^
In file included from misc2.cpp:7:
In file included from ../inst/include/pagoda2.h:15:
In file included from /Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include/space.h:32:
/Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include/utils.h:66:9: warning: 'ISNAN' macro redefined [-Wmacro-redefined]
#define ISNAN std::isnan
        ^
/Library/Frameworks/R.framework/Resources/include/R_ext/Arith.h:74:11: note: previous definition is here
#  define ISNAN(x)     R_isnancpp(x)
          ^
2 warnings generated.
/usr/local/clang4/bin/clang++ -std=gnu++11 -I/Library/Frameworks/R.framework/Resources/include -DNDEBUG  -I"/Library/Frameworks/R.framework/Versions/3.4/Resources/library/Rcpp/include" -I"/Library/Frameworks/R.framework/Versions/3.4/Resources/library/RcppArmadillo/include" -I/usr/local/include  -I"/Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include" -I"../inst/include" -I"/Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/lshkit/include" -fopenmp -fPIC  -Wall -g -O2 -c nsmlib.cpp -o nsmlib.o
In file included from nsmlib.cpp:3:
In file included from ../inst/include/pagoda2.h:15:
In file included from /Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include/space.h:32:
/Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include/utils.h:53:12: warning: 'WARN' macro redefined [-Wmacro-redefined]
#   define WARN(exp) (FILE_LINE_LINK "WARNING: " exp)
           ^
/Library/Frameworks/R.framework/Resources/include/R_ext/RS.h:60:9: note: previous definition is here
#define WARN                    WARNING(NULL)
        ^
In file included from nsmlib.cpp:3:
In file included from ../inst/include/pagoda2.h:15:
In file included from /Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include/space.h:32:
/Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include/utils.h:66:9: warning: 'ISNAN' macro redefined [-Wmacro-redefined]
#define ISNAN std::isnan
        ^
/Library/Frameworks/R.framework/Resources/include/R_ext/Arith.h:74:11: note: previous definition is here
#  define ISNAN(x)     R_isnancpp(x)
          ^
nsmlib.cpp:102:7: warning: unused variable 'nanswers' [-Wunused-variable]
  int nanswers=nqueries*k;
      ^
nsmlib.cpp:86:14: warning: variable 'index' is used uninitialized whenever 'if' condition is false [-Wsometimes-uninitialized]
  } else if (indexType == INDEX_TYPE_LP) {
             ^~~~~~~~~~~~~~~~~~~~~~~~~~
nsmlib.cpp:90:3: note: uninitialized use occurs here
  index->CreateIndex(IndexParams);
  ^~~~~
nsmlib.cpp:86:10: note: remove the 'if' if its condition is always true
  } else if (indexType == INDEX_TYPE_LP) {
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
nsmlib.cpp:81:22: note: initialize the variable 'index' to silence this warning
  Index<float> *index;
                     ^
                      = nullptr
4 warnings generated.
/usr/local/clang4/bin/clang++ -std=gnu++11 -I/Library/Frameworks/R.framework/Resources/include -DNDEBUG  -I"/Library/Frameworks/R.framework/Versions/3.4/Resources/library/Rcpp/include" -I"/Library/Frameworks/R.framework/Versions/3.4/Resources/library/RcppArmadillo/include" -I/usr/local/include  -I"/Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include" -I"../inst/include" -I"/Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/lshkit/include" -fopenmp -fPIC  -Wall -g -O2 -c pagoda1_liftover.cpp -o pagoda1_liftover.o
In file included from pagoda1_liftover.cpp:1:
In file included from ../inst/include/pagoda2.h:15:
In file included from /Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include/space.h:32:
/Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include/utils.h:53:12: warning: 'WARN' macro redefined [-Wmacro-redefined]
#   define WARN(exp) (FILE_LINE_LINK "WARNING: " exp)
           ^
/Library/Frameworks/R.framework/Resources/include/R_ext/RS.h:60:9: note: previous definition is here
#define WARN                    WARNING(NULL)
        ^
In file included from pagoda1_liftover.cpp:1:
In file included from ../inst/include/pagoda2.h:15:
In file included from /Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include/space.h:32:
/Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include/utils.h:66:9: warning: 'ISNAN' macro redefined [-Wmacro-redefined]
#define ISNAN std::isnan
        ^
/Library/Frameworks/R.framework/Resources/include/R_ext/Arith.h:74:11: note: previous definition is here
#  define ISNAN(x)     R_isnancpp(x)
          ^
2 warnings generated.
/usr/local/clang4/bin/clang++ -std=gnu++11 -I/Library/Frameworks/R.framework/Resources/include -DNDEBUG  -I"/Library/Frameworks/R.framework/Versions/3.4/Resources/library/Rcpp/include" -I"/Library/Frameworks/R.framework/Versions/3.4/Resources/library/RcppArmadillo/include" -I/usr/local/include  -I"/Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include" -I"../inst/include" -I"/Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/lshkit/include" -fopenmp -fPIC  -Wall -g -O2 -c rank.cpp -o rank.o
In file included from rank.cpp:1:
In file included from ../inst/include/pagoda2.h:15:
In file included from /Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include/space.h:32:
/Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include/utils.h:53:12: warning: 'WARN' macro redefined [-Wmacro-redefined]
#   define WARN(exp) (FILE_LINE_LINK "WARNING: " exp)
           ^
/Library/Frameworks/R.framework/Resources/include/R_ext/RS.h:60:9: note: previous definition is here
#define WARN                    WARNING(NULL)
        ^
In file included from rank.cpp:1:
In file included from ../inst/include/pagoda2.h:15:
In file included from /Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include/space.h:32:
/Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/include/utils.h:66:9: warning: 'ISNAN' macro redefined [-Wmacro-redefined]
#define ISNAN std::isnan
        ^
/Library/Frameworks/R.framework/Resources/include/R_ext/Arith.h:74:11: note: previous definition is here
#  define ISNAN(x)     R_isnancpp(x)
          ^
2 warnings generated.
/usr/local/clang4/bin/clang++ -std=gnu++11 -dynamiclib -Wl,-headerpad_max_install_names -undefined dynamic_lookup -single_module -multiply_defined suppress -L/Library/Frameworks/R.framework/Resources/lib -L/usr/local/clang4/lib -o pagoda2.so RcppExports.o binaryExport.o misc2.o nsmlib.o pagoda1_liftover.o rank.o -L/usr/lib/ -L/Users/zhengrongbin/Documents/nmslib-1.6/similarity_search/release -lNonMetricSpaceLib -lgsl -lpthread -lboost_filesystem -lboost_system -lstdc++ -L/Library/Frameworks/R.framework/Resources/lib -lRlapack -L/Library/Frameworks/R.framework/Resources/lib -lRblas -L/usr/local/gfortran/lib/gcc/x86_64-apple-darwin15/6.1.0 -L/usr/local/gfortran/lib -lgfortran -lquadmath -lm -fopenmp -F/Library/Frameworks/R.framework/.. -framework R -Wl,-framework -Wl,CoreFoundation
ld: warning: directory not found for option '-L/usr/local/gfortran/lib/gcc/x86_64-apple-darwin15/6.1.0'
installing to /Library/Frameworks/R.framework/Versions/3.4/Resources/library/pagoda2/libs
** R
** inst
** preparing package for lazy loading
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded
Error: package or namespace load failed for ‘pagoda2’ in dyn.load(file, DLLpath = DLLpath, ...):
cannot load the shared object ‘/Library/Frameworks/R.framework/Versions/3.4/Resources/library/pagoda2/libs/pagoda2.so’：:
  dlopen(/Library/Frameworks/R.framework/Versions/3.4/Resources/library/pagoda2/libs/pagoda2.so, 6): Library not loaded: libboost_filesystem.dylib
  Referenced from: /Library/Frameworks/R.framework/Versions/3.4/Resources/library/pagoda2/libs/pagoda2.so
  Reason: image not found
Error：loading failed
stop execution
ERROR: loading failed
* removing ‘/Library/Frameworks/R.framework/Versions/3.4/Resources/library/pagoda2’
Installation failed: Command failed (1)

**Comment by barkasn** [comment]
Hi,

I am afraid, I can't really help you with this as it is a system/fortran issue. I would suggest the following: make sure that fortran in correctly installed and if so try to find out why your compiler can't find it, you might be able to fix this by setting environment variables. If the issues persist you might be able to setup a symbolic link from the directory that fortran is searched for to the the directory it is actually installed.

The majority of the errors you are seeing are the result of the package not being built, the warnings are normal.

BW,

Nikolas

**Comment by barkasn** [comment]
Hi,

Have you been able to resolve this problem?

Best wishes,

Nikolas

**Comment by zhengrongbin** [comment]
sorry, I did not resolve the problem. Since I am not focusing on single-cell analysis, I have used tSNE to get my goal. But I can still find a time to have a try about your suggestion. I can imagine your tool is pretty cool.

**Comment by JiaruiMi** [comment]
Hi, I have exactly the same problem. I ended up installing the latest version of gfortran which is 6.3. My mac is Sierra 10.13.3. When I check this website http://thecoatlessprofessor.com/programming/openmp-in-r-on-os-x/#after-3-4-0, I found a note "You will need to download the OS X El Capitan gfortran 6.1 binaries regardless of whether or not you are on macOS Sierra, which presently only offers gfortran 6.3." So I guess it might be due to the gfortran version I downloaded. But unfortunately, when I uninstall the latest version of gfortran, I still cannot install the 6.1 version. The mistake message shows that there is a updated version in my disc. I actually do not know how to deal with this since I have tried all possible solution I can get to uninstall version 6.3.
